### PR TITLE
net: lwm2m: Improve thread safety of the library

### LINKED ISSUE #78989: NET_TC_THREAD_PREEMPTIVE=y results in race conditions in LwM2M engine
**Describe the bug**

Enabling the preemptive TX/RX threads introduces race conditions in the LwM2M engine.

Looking at the [socket_loop](https://github.com/zephyrproject-rtos/zephyr/blob/ac7d55b939598e702f5a118c8d31e2c0704769e8/subsys/net/lib/lwm2m/lwm2m_engine.c#L763) implementation there is almost no synchronisation. For example, scheduling a message with `lwm2m_send_cb` results in a crash due to a race condition, e.g:
```
[01:00:52.833,679] <dbg> net_lwm2m_message_handling: reply 0x20019d50 handled and removed
[01:00:52.833,831] <err> os: ***** SECURE FAULT *****
[01:00:52.833,831] <err> os:   Address: 0x10
[01:00:52.833,862] <err> os:   Attribution unit violation
[01:00:52.833,862] <err> os: r0/a1:  0x00000000  r1/a2:  0x00000000  r2/a3:  0x2001d764
[01:00:52.833,892] <err> os: r3/a4:  0x00000000 r12/ip:  0x00004000 r14/lr:  0x00015e35
[01:00:52.833,892] <err> os:  xpsr:  0x21000000
[01:00:52.833,923] <err> os: Faulting instruction address (r15/pc): 0x00044c98
[01:00:52.833,953] <err> os: >>> ZEPHYR FATAL ERROR 41: Unknown error on CPU 0
[01:00:52.833,984] <err> os: Current thread: 0x20013448 (lwm2m-sock-recv)
[01:00:52.845,520] <err> fatal_error: Resetting system
```
In this case it's a NULL-pointer dereference [here](https://github.com/zephyrproject-rtos/zephyr/blob/ac7d55b939598e702f5a118c8d31e2c0704769e8/subsys/net/lib/coap/coap.c#L1725)

**To Reproduce**

Sorry, no minimal reproducible code example. 
I think setting the `NET_TC_THREAD_PREEMPTIVE` to `y` and sending a lot of messages should suffice.

**Expected behavior**

- LwM2M-engine should be thread-safe.
- Alternatively it should not allow turning the preemptive multithreading on.

---

Sorry if this is the wrong place to post this, I just wanted to warn others of a potential problem.

The configuration flag is of course marked as experimental, so there should be no expectation that everything will work as expected, but I think since it is clear that LwM2M will not work with this flag, turning it on should result in a build error until the LwM2M engine is made thread-safe.

### PR DESCRIPTION
**Pull request by rlubos**
With preemptive network threads enabled, the LwM2M could crash when for example LwM2M send was issued from another thread. This PR addresses the potential thread-safety issues detected with the library.

Fixes #78989

### CODE REVIEW

File: subsys/net/lib/lwm2m/lwm2m_engine.h [hunk 1]
```diff
@@ -375,4 +375,19 @@ int lwm2m_sock_nfds(void);
...
+ */
+void lwm2m_client_lock(struct lwm2m_ctx *ctx);
+
+/**
+ * @brief Unlocks the client previously locked by lwm2m_client_lock().
+ *
+ * @param[in] client_ctx LwM2M context
+ */
+void lwm2m_client_unlock(struct lwm2m_ctx *ctx);
+
```
Comment (pdgendt) [inline subsys/net/lib/lwm2m/lwm2m_engine.h:391]: Nit: some extraneous lines

**Review by pdgendt — DISMISSED** [review]

File: subsys/net/lib/lwm2m/lwm2m_engine.h (hunk 1 above)
Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_engine.h:391]: Thanks, fixed

**Review by pdgendt — DISMISSED** [review]

**Review by jukkar — DISMISSED** [review]

**Review by SeppoTakalo — DISMISSED** [review]
This is acceptable, but I believe this is not all to make LwM2M thread safe.

We have also shared arrays in context
```
	struct coap_pending pendings[];
	struct coap_reply replies[];
```

Then `lwm2m_engine.c` have some global arrays that might be modified by API call from external thread
```
static struct service_node service_node_data[];
static sys_slist_t engine_service_list;
static struct zsock_pollfd sock_fds[];
static struct lwm2m_ctx *sock_ctx[];
static int sock_nfds;
static struct coap_block_context output_block_contexts[]
```

And `lwm2m_message_handling.c` also has some global arrays for messages
```
static struct lwm2m_message messages[];
static struct lwm2m_block_context block1_contexts[];
```

I did not do any analyze, if those actually are in danger, just a thought that we might need one global mutex to protect the engine as well. (And soon we have a danger of deadblock)

**Comment by rlubos** [comment]
DNM for now, I'll update the PR according to suggestions from Seppo and retest.

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c [hunk 2]
```diff
@@ -517,18 +540,25 @@ struct lwm2m_message *find_msg(struct coap_pending *pending, struct coap_reply *
 	return NULL;
 }
 
+
```
Comment (pdgendt) [inline subsys/net/lib/lwm2m/lwm2m_message_handling.c:543]: Nit: extra newline

**Review by pdgendt — DISMISSED** [review]

File: subsys/net/lib/lwm2m/lwm2m_message_handling.c (hunk 2 above)
Comment (rlubos) [reply subsys/net/lib/lwm2m/lwm2m_message_handling.c]: Thanks again, fixed

**Review by pdgendt — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Review by SeppoTakalo — APPROVED** [review]

**Pull request merged**